pkg/ring/replication_set_tracker.go

```diff
@@ -0,0 +1,104 @@
+package ring
+
+type replicationSetResultTracker interface {
+	// Signals an instance has done the execution, either successful (no error)
+	// or failed (with error).
+	done(instance *IngesterDesc, err error)
+
+	// Returns true if the minimum number of successful results have been received.
+	succeeded() bool
+
+	// Returns true if the maximum number of failed executions have been reached.
+	failed() bool
+}
+
+type defaultResultTracker struct {
+	minSucceeded int
+	numSucceeded int
+	numErrors    int
+	maxErrors    int
+}
+
+func newDefaultResultTracker(instances []IngesterDesc, maxErrors int) *defaultResultTracker {
+	return &defaultResultTracker{
+		minSucceeded: len(instances) - maxErrors,
+		numSucceeded: 0,
+		numErrors:    0,
+		maxErrors:    maxErrors,
+	}
+}
+
+func (t *defaultResultTracker) done(_ *IngesterDesc, err error) {
+	if err == nil {
+		t.numSucceeded++
+	} else {
+		t.numErrors++
+	}
+}
+
+func (t *defaultResultTracker) succeeded() bool {
+	return t.numSucceeded >= t.minSucceeded
+}
+
+func (t *defaultResultTracker) failed() bool {
+	return t.numErrors > t.maxErrors
+}
+
+type zoneAwareResultTracker struct {
```

Comment:
suggestion: I would add comment that this tracker expects that *all* instances within a zone must succeed in order for zone to succeed. (Fortunately it's quite clear from the code)

Reply:
Added